### DISCUSSION TITLE
Using reference data files in maps.

### DISCUSSION THREAD

**Discussion by dimmech** [freeplane/freeplane#1929] [General]
Is there a built in method for accomplishing the following or is it something that would be better suited for a script, addon or plugin?

I'd like to have a local directory where I could store books in the form of data files so they could be searched and referenced. This way, I could cite "book/chapter/paragraph(s)" by inserting a link on the node that would "look up" the reference and display it within the map (perhaps in a note type of dialog). I've looked through documentation and all available extendable functionality I could find but there didn't seem to be a straight forward way to accomplish this. Is that right or am I missing some basic understanding of a way to do this?

**Comment by i-plasm** [comment]
Hi @dimmech . What you describe is generally possible via extensions. What do you exactly mean by "data files", e.g what file type are they? Do they have any sort of data structure or a way to define/mark chapter and paragraph elements? Also, when you mention you'd retrieve the info via a link pointing to "book/chapter/paragraph(s)", do you have an example of what a full link might look like?

**Reply by dimmech** [reply]
Hello @i-plasm,

I'm working on a proof of concept script to convert books into structured data files like CSV, JSON, or XML. I'm also considering using a database like MongoDB. The output will be consistent because the script formats it with null values for any criteria that doesn't match (as in the example below). The structure includes values for Book/Section/Chapter/Subchapter/Paragraph. If a section does not exist, a null entry is used. I'm continually tweaking the script to accommodate variations in the formatting of source books.

I'd like to add links such as "Don Quixote/Chapter 3/Paragraph 1" for a connected DB or reference a file as "Path/to/Don_Quixote.json/Chapter 3/Paragraph 1". This way, I can cite "book/chapter/paragraph(s)" by inserting a link on the node that would look up the reference and display it within the map (perhaps in a note dialog).

Here's an example of the JSON structured output I'm using:

```json
{
  "book": {
    "title": "Don Quixote",
    "author": "Miguel de Cervantes",
    "sections": [
      {
        "section_number": 1,
        "section_title": "Part 1",
        "chapters": [
          {
            "chapter_number": 1,
            "chapter_title": "Which treats of the character and pursuits of the famous gentleman Don Quixote of La Mancha",
            "subchapters": [
              {
                "subchapter_number": null,
                "subchapter_title": null,
                "paragraphs": [
                  {
                    "paragraph_number": 1,
                    "content": "In a village of La Mancha, the name of which I have no desire to call to mind, there lived not long since one of those gentlemen that keep a lance in the lance-rack, an old buckler, a lean hack, and a greyhound for coursing..."
                  }
                ]
              }
            ]
          },
          {
            "chapter_number": 2,
            "chapter_title": "Which treats of the first sally the ingenious Don Quixote made from home",
            "subchapters": [
              {
                "subchapter_number": null,
                "subchapter_title": null,
                "paragraphs": [
                  {
                    "paragraph_number": 1,
                    "content": "So saying, he slipped on his armor, mounted Rocinante with his patched-up helmet on, and, grasping his lance, set off along the highroad as proud and satisfied as if he were taking part in a tournament..."
                  }
                ]
              }
            ]
          }
          // Additional chapters...
        ]
      }
    ]
  }
}
```

Would this approach be feasible with Freeplane extensions, or would you recommend a different method to achieve this functionality? Additionally, any tips on improving this concept would be greatly appreciated.

**Reply by i-plasm** [reply]
Hello @dimmech ,

The ability to quickly reference and retrieve book portions sounds quite practical indeed.

One approach is to have Freeplane be in charge of the tasks of resolving the custom hyperlink and parsing it for the required paragraph(s). 

A more "global" approach would involve a web server that would resolve an http URL and return the requested paragraphs. This last approach is obviously required if you need hyperlinks to work globally (e.g in Microsoft Office, browsers, etc), but it seems like too much if you don't really need it. 

For the first approach, a Freeplane extension would perform the following tasks:

1) Detect hyperlinks -let's say on mouse hover- structured like *"Don Quixote/Chapter 3/Paragraph 1"*
   - Note 1: Detection on Mouse hover would allow the placement of hyperlinks virtually anywhere in freeplane - node link and hyperlinks inside the text of node core/note/details
   - Note 2: This type of hyperlink would be portable between computers, so I think that's the most convenient one. Additionally, it allows decoupling from the file data type/data source, so you are free to make design changes in the future and your map links will remain valid. Of course, this requires some sort of preferences file for specifying e.g books folder root directory, data file type/extension, etc.

2) Map and resolve the hyperlink to the actual JSON file

3) Parse the JSON and extract the target paragraph(s), converting this into formatted html if desired

4) Launch a popup viewer displaying the plain text/html

If you want the book fragments to be actually extracted into node core/notes/etc -so that they become searchable when using Freeplane search engines-, a variation would be able to do so. Last, Json seems like a good choice. I think it would be better to create Json files per chapter and not per book, for performance reasons, as the parsing of long enough books might take too long.

**Comment by macmarrum** [comment]
### Option 1

You can write an extension for Freeplane (a Groovy script) that will use your reference from a node in Freeplane, look it up in your database, find the text you want and display it in Freeplane (dialog window).

### Option 2

Freeplane has the functionality to execute any os command (using Groovy's `[command, arg1, arg2].execute()` or via `Insert->Link->Add hyperlink to execution of an OS command…`). The command can include parameters, like your "Don Quixote/Chapter 3/Paragraph 1".

There's also a way to trigger a Groovy script in Freeplane from outside of Freeplane (aka remote control).

#### Step-by-step example

Let's say you have a script `reference_lookup.py` that can receive your reference, look it up in your database and find the text that you want inserted in Freeplane on a node (e.g. as its note).

```groovy
// This is a generic Freeplane script that will run the command
// with the parameter sourced from the selected node's text
// It can be saved `<user-dir>/scripts/myScriptName.groovy`
// and bound to a hot key
['reference_lookup.py', node.text].execute()
```

Using [init/remoteControl.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/init/remoteControl.groovy) and [freeplane_remote.py](https://github.com/macmarrum/freeplane-scripting/blob/main/src/python/freeplane_remote.py), `reference_lookup.py` can send a command back to Freeplane to display the text it found earlier.

```python
import sys
from pathlib import Path
from freeplane_remote import transfer

reference = sys.argv[1]

def lookup(reference):
    # your code to look the text in your DB based on the reference
    ...
    return text

text = lookup(reference)

# save it to a file
Path('/tmp/reference_lookup.txt').write_text(text, encoding='UTF-8')

groovy_script = """\
// read the text of the file to the current node's note
def tempFile = new File('/tmp/reference_lookup.text')
node.note = tempFile.get_text('UTF-8')
// open a dialog to view/edit the note
menuUtils.executeMenuItems(['EditNoteInDialogAction'])
// delete the no-longer needed file
tempFile.delete()
"""

transfer(groovy_script)
```

**Note**: _you might need to quote the tmp-file path – see `quote_path` in https://github.com/macmarrum/freeplane-scripting/blob/main/src/python/freeplane-remote-diff.py_

**Comment by dimmech** [comment]
@macmarrum and @i-plasm,

Thank you both for your insightful suggestions and detailed explanations. Your expertise is greatly appreciated in helping me understand what my commitment would entail and what course of action would be best if I try to implement this functionality in Freeplane.

Based on our discussions so far, I agree that a local solution is a good starting point, but I also find a global DB solution very attractive.

Assuming one is brand new to Groovy as well as Freeplane's practices, methodologies, and API, please advise me on some conceptual areas of focus and any reference materials that might help me accomplish this. Additionally, any extensions or code written that provide insight into any portion of the goal would be invaluable. I've started to dive a little deeper into the documentation and am still evaluating if I have room on my plate for this.

One thing I'm thinking about is that I would have to remember the variations of every book's structure ahead of time in order to write hyperlinks effectively. It seems a "reference selection" method would be more appropriate, given a suitable way to select the paragraph(s), which could probably be done through a search or cascading dropdown boxes. Ideally, an interactive selection tool would be the cat's meow; however, I don't want to get ahead of myself.

Either way, I'm willing to see where this leads and welcome any input.

**Reply by macmarrum** [reply]
> Assuming one is brand new to Groovy as well as Freeplane's practices, methodologies and api, please advise me on some conceptual areas of focus and any reference material that might help me accomplish this.

Without prior knowledge of Groovy or Java, this might be a truly challenging task. I'd recommend going with a programming language you're already comfortable with.

If you're set on going with Groovy, I'd start with https://docs.freeplane.org/scripting/Learning_guide.html

> I've started to dive a little deeper into the documentation and am still evaluating if I have room on my plate for this.

You might want to have a look at http://docear.org to see if it covers (some of) your needs.

**Reply by dimmech** [reply]
Thanks for the link to Docear, @macmarrum. I haven't seen that before. While it looks interesting, I don't think it's the solution I'd personally use since I've been invested in creating maps with Freeplane for a few years now on just about everything, and a Freeplane solution would just be awesome.

I appreciate your honest assessment of the challenges involved in learning Groovy from scratch. However, I've got a little Java experience from "back in the day" (it's been some time...) and didn't realize that Groovy was a simplified version of Java until now. I might just go ahead and try to evolve a little by taking this on. I think that with the docs and occasional community support, I can cobble something together. Let's see what happens!

If anyone has additional tips or experiences with Freeplane development or API that would prove useful to this particular concept, I'd love to hear them. Specifically, any advice on how to implement a "reference selection" method or interactive tool for selecting paragraphs and formatting links based on the selection would be incredibly helpful.

Thanks again for your support!

**Reply by macmarrum** [reply]
> Thanks for the link to Docear, @macmarrum. I haven't seen that before. While it looks interesting, I don't think it's the solution I'd personally use since I've been invested in creating maps with Freeplane for a few years now on just about everything, and a Freeplane solution would just be awesome.

Interestingly, Docear is based on Freeplane 1.3.2 alpha.

> I appreciate your honest assessment of the challenges involved in learning Groovy from scratch. However, I've got a little Java experience from "back in the day" (it's been some time...) and didn't realize that Groovy was a simplified version of Java until now. I might just go ahead and try to evolve a little by taking this on. I think that with the docs and occasional community support, I can cobble something together. Let's see what happens!

:+1:

> If anyone has additional tips or experiences with Freeplane development or API that would prove useful to this particular concept, I'd love to hear them. Specifically, any advice on how to implement a "reference selection" method or interactive tool for selecting paragraphs and formatting links based on the selection would be incredibly helpful.

The way I see it, there's very little that you need to know about Freeplane development or API. 99% or the work will be outside of Freeplane. You might want to use Java Swing, since it's what Freeplane uses for its UI. I can help with the remaining 1% to integrate your code with Freeplane, e.g.

```groovy
// I assume your application (DimmechReferenceSelector) will provide a dialog window
// that will allow to search for or select a reference, and the corresponding method
// (e.g. showSelectionDialog) will return the selected reference
def app = new DimmechReferenceSelector()
def reference = app.showSelectionDialog()
if (reference) {
  // replace the selected node's core (text) with the reference
  node.text = reference
}
```
The above script can be assigned a hot key (e.g. `F9`) or be launched from the context menu (right click on a node).

It's worth to note that you can write your application entirely in Java, create a jar and provide it to Freeplane so that it's available for scripting. Alternatively, your application can be a Groovy library in the form of a .groovy file, which Freeplane will compile for you during start-up.
